File: custom-rings/sdk/src/velocity.rs
//! The record slots a velocity transfer adds and the witness the circuit charges them with.

use rand::{rngs::OsRng, RngCore};
use solana_address::Address;
use zolana_client::Shape;
use zolana_interface::instruction::MessageData;
use zolana_keypair::{NullifierKey, PublicKey, ShieldedAddress, ViewingKey, ViewingKeyTrait};
use zolana_ring_policy::{
    ring_id_field, ListNamespace, Member, SpendCounters, SpendRecord, VelocityRow,
    MAX_VELOCITY_ASSETS,
};
use zolana_transaction::{
    instructions::{transact::SppProofOutputUtxo, types::SppProofInputUtxo},
    utxo::derive_transact_output_blinding,
    Data, Utxo, SOL_MINT,
};

use zolana_ring_client::{
    counters_message, decrypt_counters, encrypt_counters, find_counters_message,
};

use crate::{
    instructions::spend::LiveSpendRecord,
    instructions::transact::{SpendRecordWitness, VelocityWitness},
    TransferError,
};

pub(crate) struct VelocityFacts {
    pub namespace: Address,
    pub owner: ListNamespace,
    pub entries_tree_id: u16,
    pub window_slots: u64,
    pub rows: Vec<VelocityRow>,
    pub window_index: u64,
    pub live: LiveSpendRecord,
    /// `None` for an expired record, the circuit opens only its commitment.
    pub counters: Option<SpendCounters>,
}

pub(crate) struct VelocityContext<'a> {
    pub namespace: Address,
    pub owner: ListNamespace,
    pub entries_tree_id: u16,
    pub window_slots: u64,
    pub rows: Vec<VelocityRow>,
    pub sender: &'a (dyn ViewingKeyTrait + Send + Sync),
}

impl VelocityContext<'_> {
    pub(crate) fn facts(
        self,
        live: LiveSpendRecord,
        slot: u64,
    ) -> Result<VelocityFacts, TransferError> {
        let window_index = slot / self.window_slots;
        let counters = VelocityFacts::recover_counters(
            &live,
            window_index,
            self.namespace.as_array(),
            self.sender,
        )?;
        Ok(VelocityFacts {
            namespace: self.namespace,
            owner: self.owner,
            entries_tree_id: self.entries_tree_id,
            window_slots: self.window_slots,
            rows: self.rows,
            window_index,
            live,
            counters,
        })
    }
}

impl VelocityFacts {
    pub(crate) fn recover_counters(
        live: &LiveSpendRecord,
        window_index: u64,
        namespace: &[u8; 32],
        sender: &(dyn ViewingKeyTrait + Send + Sync),
    ) -> Result<Option<SpendCounters>, TransferError> {
        if live.record.window != window_index {
            return Ok(None);
        }
        let counters = if live.record.version == 0 {
            SpendCounters::zero(&[])
        } else {
            let message = find_counters_message(&live.origin.messages, namespace)
                .ok_or(TransferError::SpendCountersUnknown)?;
            let salt = live
                .origin
                .salt
                .ok_or(TransferError::SpendCountersUnknown)?;
            let tx_key = sender.get_transaction_viewing_key(&live.origin.first_nullifier)?;
            decrypt_counters(&tx_key, &message.data, salt)
                .map_err(|_| TransferError::SpendCountersUnknown)?
        };
        if counters
            .commitment()
            .map_err(|_| TransferError::PolicyHashing)?
            != live.record.counters_commitment
        {
            return Err(TransferError::SpendCountersUnknown);
        }
        Ok(Some(counters))
    }
}

pub(crate) struct Outflows<'a> {
    pub sender: Member,
    pub ring: Address,
    pub inputs: &'a [SppProofInputUtxo],
    pub outputs: &'a [SppProofOutputUtxo],
}

impl Outflows<'_> {
    /// Inputs of the mint less the sender's change inside the ring, as the circuit sums them.
    fn outflow(&self, asset: &[u8; 32]) -> Result<u64, TransferError> {
        let mut inflow: u64 = 0;
        for input in self.inputs.iter().filter(|input| !input.is_dummy()) {
            if Member::asset(&input.utxo.asset)
                .map_err(|_| TransferError::PolicyHashing)?
                .as_bytes()
                == asset
            {
                inflow = inflow
                    .checked_add(input.utxo.amount)
                    .ok_or(TransferError::VelocityOverflow)?;
            }
        }
        let mut change: u64 = 0;
        for output in self.outputs {
            let Some(address) = output.owner_address.as_ref() else {
                continue;
            };
            let owner = address
                .signing_pubkey
                .owner_proof_input_hash()
                .map_err(|_| TransferError::PolicyHashing)?;
            let same_asset = Member::asset(&output.asset)
                .map_err(|_| TransferError::PolicyHashing)?
                .as_bytes()
                == asset;
            if same_asset
                && owner == *self.sender.as_bytes()
                && output.ring_program_id == Some(self.ring)
            {
                change = change
                    .checked_add(output.amount)
                    .ok_or(TransferError::VelocityOverflow)?;
            }
        }
        inflow
            .checked_sub(change)
            .ok_or(TransferError::VelocityOverflow)
    }
}

/// The rows charged for one transfer, shared by both velocity modes.
#[derive(Debug)]
pub(crate) struct RowCharges {
    pub rows: [VelocityRow; MAX_VELOCITY_ASSETS],
    pub row_count: u8,
    pub spent: [u64; MAX_VELOCITY_ASSETS],
    pub approval_required: bool,
}

/// Charges each row's outflow, `previous` supplied only inside a live window.
pub(crate) struct ChargeRows<'a> {
    pub rows: &'a [VelocityRow],
    pub outflows: &'a Outflows<'a>,
    pub previous: Option<&'a SpendCounters>,
}

impl ChargeRows<'_> {
    pub(crate) fn charge(self) -> Result<RowCharges, TransferError> {
        let mut charges = RowCharges {
            rows: [VelocityRow {
                asset: [0u8; 32],
                cap: 0,
                cosign_above: 0,
            }; MAX_VELOCITY_ASSETS],
            row_count: self.rows.len() as u8,
            spent: [0u64; MAX_VELOCITY_ASSETS],
            approval_required: false,
        };
        for (index, row) in self.rows.iter().enumerate() {
            charges.rows[index] = *row;
            let outflow = self.outflows.outflow(&row.asset)?;
            let previous = self
                .previous
                .map_or(0, |counters| counters.spent(&row.asset));
            let spent = previous
                .checked_add(outflow)
                .ok_or(TransferError::VelocityOverflow)?;
            if row.cap != 0 && spent > row.cap {
                return Err(TransferError::VelocityCapExceeded {
                    asset: row.asset,
                    cap: row.cap,
                    spent,
                });
            }
            charges.approval_required |= row.cosign_above != 0 && outflow > row.cosign_above;
            charges.spent[index] = spent;
        }
        Ok(charges)
    }
}

/// A field element below the modulus, the Poseidon commitment rejects the rest.
fn canonical_salt() -> [u8; 32] {
    let mut salt = [0u8; 32];
    OsRng.fill_bytes(&mut salt);
    salt[0] = 0;
    salt
}

pub(crate) struct VelocityPlan {
    pub input: SppProofInputUtxo,
    pub output: SppProofOutputUtxo,
    /// The plaintext record slot `finalize` publishes at the output's position.
    pub record_slot: MessageData,
    pub counters_message: MessageData,
    pub witness: VelocityWitness,
    pub shape: Shape,
}

pub(crate) struct VelocityPlanInput<'a> {
    pub facts: &'a VelocityFacts,
    pub outflows: Outflows<'a>,
    pub tx_viewing_key: &'a ViewingKey,
    pub salt: [u8; 16],
    pub first_nullifier: [u8; 32],
    pub output_blinding_seed: [u8; 32],
    /// The money slots the record follows, dummies included.
    pub money_shape: Shape,
}

impl VelocityPlanInput<'_> {
    pub(crate) fn plan(self) -> Result<VelocityPlan, TransferError> {
        let facts = self.facts;
        let shape = record_shape(self.money_shape)?;
        let same_window = facts.live.record.window == facts.window_index;
        let previous = facts.counters.as_ref().filter(|_| same_window);
        let charges = ChargeRows {
            rows: &facts.rows,
            outflows: &self.outflows,
            previous,
        }
        .charge()?;
        let mut next = SpendCounters::zero(&[]);
        next.salt = canonical_salt();
        for index in 0..usize::from(charges.row_count) {
            next.assets[index] = charges.rows[index].asset;
            next.spent[index] = charges.spent[index];
        }
        let commitment = next
            .commitment()
            .map_err(|_| TransferError::PolicyHashing)?;
        let spent = &facts.live.record;
        let successor = SpendRecord {
            member: spent.member,
            version: spent
                .version
                .checked_add(1)
                .ok_or(TransferError::VelocityOverflow)?,
            window: facts.window_index,
            counters_commitment: commitment,
            blinding: derive_transact_output_blinding(
                &self.first_nullifier,
                &self.output_blinding_seed,
                shape.n_outputs() as u32 - 1,
            )?,
        };
        let address = facts
            .owner
            .spend_address(&spent.member, facts.entries_tree_id)
            .map_err(|_| TransferError::PolicyHashing)?;
        let spent_data_hash = spent
            .data_hash(&address)
            .map_err(|_| TransferError::PolicyHashing)?;
        let next_data_hash = successor
            .data_hash(&address)
            .map_err(|_| TransferError::PolicyHashing)?;
        let namespace_owner = PublicKey::from_pda(&facts.namespace);
        let zero_nullifier = NullifierKey::from_secret([0u8; 31]);
        let input = SppProofInputUtxo {
            utxo: Utxo {
                owner: namespace_owner,
                asset: SOL_MINT,
                amount: 0,
                blinding: spent.blinding,
                ring_program_id: None,
                data: Data::default(),
            },
            nullifier_key: zero_nullifier.clone(),
            data_hash: Some(spent_data_hash),
            ring_data_hash: None,
            tree_id: facts.entries_tree_id,
        };
        let output = SppProofOutputUtxo {
            asset: SOL_MINT,
            amount: 0,
            blinding: successor.blinding,
            ring_program_id: None,
            ring_data_hash: None,
            data_hash: Some(next_data_hash),
            owner_address: Some(ShieldedAddress::for_pda(
                &facts.namespace,
                zero_nullifier.pubkey()?,
                ViewingKey::new().pubkey(),
            )),
            owner_tag: Some(facts.namespace.to_bytes()),
            data: Data::default(),
        };
        let counters_body = encrypt_counters(
            self.tx_viewing_key,
            &self.tx_viewing_key.pubkey(),
            self.salt,
            &next,
        )?;
        let opened = facts.counters.unwrap_or_else(|| SpendCounters::zero(&[]));
        let witness = VelocityWitness {
            window_slots: facts.window_slots,
            rows: charges.rows,
            row_count: charges.row_count,
            ring_id: ring_id_field(self.outflows.ring.as_array())
                .map_err(|_| TransferError::PolicyHashing)?,
            namespace_owner_hash: facts.owner.owner_hash,
            window_index: facts.window_index,
            approval_required: charges.approval_required,
            record: SpendRecordWitness {
                version: spent.version,
                window: spent.window,
                commitment: spent.counters_commitment,
                salt: opened.salt,
                assets: opened.assets,
                spent: opened.spent,
                next_salt: next.salt,
            },
        };
        Ok(VelocityPlan {
            input,
            output,
            record_slot: MessageData {
                view_tag: facts.namespace.to_bytes(),
                data: successor.to_output_data().to_vec(),
            },
            counters_message: counters_message(facts.namespace.to_bytes(), counters_body),
            witness,
            shape,
        })
    }
}

/// The smallest supported shape with one slot beyond the money on each side.
pub(crate) fn record_shape(money: Shape) -> Result<Shape, TransferError> {
    let n_in = money.n_inputs() + 1;
    let n_out = money.n_outputs() + 1;
    zolana_client::SPP_SUPPORTED_SHAPES
        .into_iter()
        .filter(|shape| {
            shape.n_inputs() <= zolana_ring_policy::POLICY_INPUT_SLOTS
                && shape.n_outputs() <= zolana_ring_policy::POLICY_OUTPUT_SLOTS
        })
        .find(|shape| shape.n_inputs() >= n_in && shape.n_outputs() >= n_out)
        .ok_or(TransferError::PolicyShapeUnsupported)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn mint() -> Address {
        Address::new_from_array([9u8; 32])
    }

    fn row_asset() -> [u8; 32] {
        *Member::asset(&mint()).expect("asset member").as_bytes()
    }

    fn money_input(amount: u64) -> SppProofInputUtxo {
        SppProofInputUtxo {
            utxo: Utxo {
                owner: PublicKey::from_pda(&Address::new_from_array([7u8; 32])),
                asset: mint(),
                amount,
                blinding: [0u8; 32],
                ring_program_id: None,
                data: Data::default(),
            },
            nullifier_key: NullifierKey::from_secret([0u8; 31]),
            data_hash: None,
            ring_data_hash: None,
            tree_id: 0,
        }
    }

    fn charge(
        amount: u64,
        cap: u64,
        cosign_above: u64,
        previous: Option<u64>,
    ) -> Result<RowCharges, TransferError> {
        let inputs = [money_input(amount)];
        let outputs: [SppProofOutputUtxo; 0] = [];
        let outflows = Outflows {
            sender: Member::owner_tag(&[1u8; 32]).expect("sender"),
            ring: Address::default(),
            inputs: &inputs,
            outputs: &outputs,
        };
        let row = VelocityRow {
            asset: row_asset(),
            cap,
            cosign_above,
        };
        let counters = previous.map(|spent| {
            let mut counters = SpendCounters::zero(&[row_asset()]);
            counters.spent[0] = spent;
            counters
        });
        ChargeRows {
            rows: &[row],
            outflows: &outflows,
            previous: counters.as_ref(),
        }
        .charge()
    }

    #[test]
    fn the_record_takes_the_slot_after_the_money() {
        assert_eq!(record_shape(Shape::IN1_OUT1).unwrap(), Shape::IN2_OUT2);
        assert_eq!(record_shape(Shape::IN1_OUT2).unwrap(), Shape::IN2_OUT3);
        assert_eq!(record_shape(Shape::IN2_OUT3).unwrap(), Shape::IN4_OUT4);
        assert_eq!(record_shape(Shape::IN4_OUT3).unwrap(), Shape::IN5_OUT4);
        assert!(record_shape(Shape::IN4_OUT4).is_err());
    }

    #[test]
    fn a_transfer_under_the_cap_charges_its_outflow() {
        let charges = charge(100, 1000, 0, None).expect("under the cap");
        assert_eq!(charges.spent[0], 100);
        assert!(!charges.approval_required);
    }

    #[test]
    fn a_transfer_at_the_cap_passes() {
        let charges = charge(1000, 1000, 0, None).expect("at the cap");
        assert_eq!(charges.spent[0], 1000);
    }

    #[test]
    fn a_transfer_over_the_cap_is_refused_with_its_outflow() {
        let error = charge(1001, 1000, 0, None).expect_err("over the cap");
        assert!(matches!(
            error,
            TransferError::VelocityCapExceeded {
                asset,
                cap: 1000,
                spent: 1001,
            } if asset == row_asset()
        ));
    }

    #[test]
    fn the_threshold_reads_the_outflow_alone() {
        assert!(charge(600, 0, 500, None).expect("above").approval_required);
        assert!(!charge(500, 0, 500, None).expect("at").approval_required);
    }

    #[test]
    fn a_supplied_previous_adds_to_the_charge() {
        assert_eq!(charge(100, 0, 0, Some(50)).expect("windowed").spent[0], 150);
        assert_eq!(charge(100, 0, 0, None).expect("per transfer").spent[0], 100);
    }

    #[test]
    fn a_previous_over_the_cap_is_refused() {
        let error = charge(600, 1000, 0, Some(500)).expect_err("cumulative over the cap");
        assert!(matches!(
            error,
            TransferError::VelocityCapExceeded { spent: 1100, .. }
        ));
    }

    #[test]
    fn the_salt_stays_below_the_modulus() {
        for _ in 0..1000 {
            assert_eq!(canonical_salt()[0], 0);
        }
    }

    #[test]
    fn a_per_transfer_witness_carries_the_rows_and_no_record() {
        let charges = charge(100, 1000, 0, None).expect("charges");
        let witness = VelocityWitness::per_transfer(&charges, [1u8; 32], [2u8; 32]);
        assert_eq!(witness.window_slots, 0);
        assert_eq!(witness.window_index, 0);
        assert_eq!(witness.row_count, 1);
        assert_eq!(witness.record, SpendRecordWitness::default());
    }
}

File: sdk-libs/ts/src/ring/velocity.ts
import type { Address } from "@solana/kit";
import { NullifierKey } from "../keypair/nullifier-key.js";
import { ShieldedAddress } from "../keypair/shielded.js";
import { ShieldedPublicKey } from "../keypair/public-key.js";
import { ViewingKey } from "../keypair/viewing-key.js";
import { randomBlinding } from "../keypair/bytes.js";
import { transactOutputBlinding } from "../keypair/transact/index.js";
import { hashBytes } from "../hasher/index.js";
import type { Shape } from "../interface/shape.js";
import { SPP_SUPPORTED_SHAPES } from "../interface/shape.js";
import type { Bytes31, Bytes32, MessageData, RequestContext } from "../interface/types.js";
import { Utxo, ProofInputUtxo, createProofOutput } from "../transaction/utxo.js";
import type { ProofOutputUtxo, TreeId } from "../transaction/utxo.js";
import { SOL_MINT } from "../transaction/asset.js";
import { U64_MAX, decodeAddress } from "../transaction/internal.js";
import { equalBytes } from "../wallet/internal.js";
import type { SealedMessageInput, SpendSession } from "../transaction/wallet/authority.js";
import type {
  SlotReader,
  ChainReader,
  RingHeadReader,
  RingHeadTransferProof,
} from "../client/ports.js";
import {
  RING_INPUT_SLOTS,
  RING_OUTPUT_SLOTS,
  velocityProofInputOff,
} from "../client/prover/types.js";
import type {
  CustomRingVelocityRow,
  CustomRingVelocityProofInput,
} from "../client/prover/types.js";
import { RingError } from "./error.js";
import {
  RingListNamespace,
  memberOfAsset,
  memberOfIdentity,
  ringNamespaceOwnerHash,
  spendCountersCommitment,
  spendCountersSpent,
  zeroSpendCounters,
  encodeSpendRecord,
  spendRecordMessageTag,
  type LiveSpendRecord,
  type Member,
  type SpendCounters,
  type SpendRecord,
} from "./policy.js";
import { findSpendCountersMessage, openSpendCounters, sealedSpendCounters } from "./counters.js";
import { readCurrentSpendRecord } from "./head-reader.js";

const ZERO_NULLIFIER_SECRET = new Uint8Array(31) as Bytes31;

/** Excludes the record slot from money openings for outflow accounting. */
export interface RingMovement {
  readonly sender: Member;
  readonly ringProgramId: Address;
  readonly inputs: readonly ProofInputUtxo[];
  readonly outputs: readonly ProofOutputUtxo[];
}

/** Opens authenticated counters only within their record's window. */
export interface VelocityFacts {
  readonly namespace: Address;
  readonly owner: RingListNamespace;
  readonly entriesTreeId: TreeId;
  readonly windowSlots: bigint;
  readonly rows: readonly CustomRingVelocityRow[];
  readonly windowIndex: bigint;
  readonly live: LiveSpendRecord;
  /** `undefined` for an expired record, the circuit opens only its commitment. */
  readonly counters: SpendCounters | undefined;
  readonly head: RingHeadTransferProof;
}

/** Locates the sender's compressed record under the configured entries tree. */
export interface ReadVelocityFactsInput {
  readonly client: Pick<RingHeadReader, "getRingHeadTransferProof"> &
    SlotReader &
    Pick<ChainReader, "getAccount">;
  readonly ringProgramId: Address;
  readonly session: Pick<SpendSession, "openSealedMessage">;
  readonly namespace: Address;
  readonly entriesTree: Address;
  readonly entriesTreeId: TreeId;
  readonly windowSlots: bigint;
  readonly rows: readonly CustomRingVelocityRow[];
  readonly sender: Member;
}

/** Couples the successor counters to the record spend and head transition. */
export interface VelocityPlan {
  readonly nextNullifier: Bytes32;
  readonly shape: Shape;
  readonly recordInput: ProofInputUtxo;
  readonly recordOutput: ProofOutputUtxo;
  readonly recordMessage: MessageData;
  readonly countersSeal: Omit<SealedMessageInput, "slotIndex">;
  readonly proofInput: CustomRingVelocityProofInput;
  readonly approvalRequired: boolean;
}

/** Binds record accounting to the money transfer and its output blindings. */
export interface PlanVelocityInput {
  readonly facts: VelocityFacts;
  readonly movement: RingMovement;
  readonly firstNullifier: Bytes32;
  readonly outputBlindingSeed: Bytes32;
  readonly moneyShape: Shape;
}

/** The smallest supported shape with one slot beyond the money on each side. */
export function recordShape(money: Shape): Shape {
  const shape = SPP_SUPPORTED_SHAPES.find(
    (candidate) =>
      candidate.inputs <= RING_INPUT_SLOTS &&
      candidate.outputs <= RING_OUTPUT_SLOTS &&
      candidate.inputs >= money.inputs + 1 &&
      candidate.outputs >= money.outputs + 1,
  );
  if (shape === undefined) {
    throw new RingError("RING_BUILD_TRANSFER", {
      details: { reason: "recordShape" },
    });
  }
  return shape;
}

/** The sender's outflow of one mint, inputs less the change kept inside the ring. */
export function senderOutflow(movement: RingMovement, asset: Bytes32): bigint {
  let inflow = 0n;
  for (const input of movement.inputs) {
    if (input.isDummy()) continue;
    if (equalBytes(memberOfAsset(input.utxo.asset), asset)) inflow += input.utxo.amount;
  }
  let change = 0n;
  for (const output of movement.outputs) {
    const owner = output.ownerAddress;
    if (owner === undefined) continue;
    const sameAsset = equalBytes(memberOfAsset(output.asset), asset);
    const sameOwner = equalBytes(
      memberOfIdentity(owner.signingPublicKey.ownerProofInputHash()),
      movement.sender,
    );
    const inRing =
      output.ringProgramId !== undefined && output.ringProgramId === movement.ringProgramId;
    if (sameAsset && sameOwner && inRing) change += output.amount;
  }
  if (change > inflow) throw new RingError("RING_VELOCITY_OVERFLOW", { details: { asset } });
  const outflow = inflow - change;
  if (outflow > U64_MAX) throw new RingError("RING_VELOCITY_OVERFLOW", { details: { asset } });
  return outflow;
}

export async function readVelocityFacts(
  input: ReadVelocityFactsInput,
  context?: RequestContext,
): Promise<VelocityFacts> {
  const owner = RingListNamespace.of(input.namespace, input.entriesTreeId);
  if (input.windowSlots <= 0n) throw new RingError("RING_VELOCITY_DISABLED");
  const { head, live } = await readCurrentSpendRecord(input, context);
  const slot = await input.client.getSlot(context);
  const windowIndex = slot / input.windowSlots;
  const counters = await recoverCounters(input, live, windowIndex);
  return Object.freeze({
    namespace: input.namespace,
    owner,
    entriesTreeId: input.entriesTreeId,
    windowSlots: input.windowSlots,
    rows: input.rows,
    windowIndex,
    live,
    counters,
    head,
  });
}

async function recoverCounters(
  input: ReadVelocityFactsInput,
  live: LiveSpendRecord,
  windowIndex: bigint,
): Promise<SpendCounters | undefined> {
  if (live.record.window > windowIndex) throw new RingError("RING_SPEND_RECORD_INVALID");
  if (live.record.window < windowIndex) return undefined;
  if (live.record.version === 0n) return zeroSpendCounters();
  const message = findSpendCountersMessage(live.origin.messages, decodeAddress(input.namespace));
  if (message === undefined || live.origin.salt === undefined) {
    throw new RingError("RING_SPEND_COUNTERS_UNKNOWN", {
      details: { reason: message === undefined ? "missingMessage" : "missingSalt" },
    });
  }
  return openSpendCounters(input.session, {
    firstNullifier: live.origin.firstNullifier,
    salt: live.origin.salt,
    data: message.data,
    commitment: live.record.countersCommitment,
  });
}

export function planVelocity(input: PlanVelocityInput): VelocityPlan {
  // 1. Charge outflow against counters from the current window only.
  const { facts, movement } = input;
  const shape = recordShape(input.moneyShape);
  const sameWindow = facts.live.record.window === facts.windowIndex;
  const previous = sameWindow ? facts.counters : undefined;

  const rows: CustomRingVelocityRow[] = [];
  const spent: bigint[] = [];
  let approvalRequired = false;
  for (const row of facts.rows) {
    const outflow = senderOutflow(movement, row.asset);
    const before = previous === undefined ? 0n : spendCountersSpent(previous, row.asset);
    const charged = before + outflow;
    if (charged > U64_MAX) {
      throw new RingError("RING_VELOCITY_OVERFLOW", { details: { asset: row.asset } });
    }
    if (row.cap !== 0n && charged > row.cap) {
      throw new RingError("RING_VELOCITY_CAP_EXCEEDED", {
        details: { asset: row.asset, cap: row.cap, spent: charged },
      });
    }
    if (row.cosignAbove !== 0n && outflow > row.cosignAbove) approvalRequired = true;
    rows.push(row);
    spent.push(charged);
  }

  // 2. Bind successor counters to fresh salt and the SPP output blinding.
  const nextSalt = randomBlinding();
  const nextCounters: SpendCounters = Object.freeze({
    salt: nextSalt,
    assets: Object.freeze(rows.map((row) => row.asset)),
    spent: Object.freeze([...spent]),
  });
  const commitment = spendCountersCommitment(nextCounters);

  const spentRecord = facts.live.record;
  const successor: SpendRecord = Object.freeze({
    member: spentRecord.member,
    version: spentRecord.version + 1n,
    window: facts.windowIndex,
    countersCommitment: commitment,
    blinding: transactOutputBlinding(
      input.firstNullifier,
      input.outputBlindingSeed,
      shape.outputs - 1,
    ),
  });

  // 3. Keep the namespace-owned record outside money subjects.
  const namespace = decodeAddress(facts.namespace);
  const spentHashes = facts.owner.spendRecordHashes(spentRecord);
  const nextHashes = facts.owner.spendRecordHashes(successor);
  const zeroNullifier = NullifierKey.fromSecret(ZERO_NULLIFIER_SECRET);
  const viewing = ViewingKey.generate();
  let recordInput: ProofInputUtxo;
  let recordOutput: ProofOutputUtxo;
  try {
    recordInput = new ProofInputUtxo({
      utxo: new Utxo({
        owner: ShieldedPublicKey.fromPda(namespace),
        asset: SOL_MINT,
        amount: 0n,
        blinding: spentRecord.blinding,
      }),
      nullifierKey: zeroNullifier,
      treeId: facts.entriesTreeId,
      dataHash: spentHashes.dataHash,
    });
    recordOutput = createProofOutput({
      asset: SOL_MINT,
      amount: 0n,
      blinding: successor.blinding,
      dataHash: nextHashes.dataHash,
      ownerAddress: ShieldedAddress.forPda({
        pda: namespace,
        nullifierPublicKey: zeroNullifier.publicKey(),
        viewingPublicKey: viewing.publicKey(),
      }),
      ownerTag: namespace,
    });
  } finally {
    zeroNullifier.destroy();
    viewing.destroy();
  }

  const opened = facts.counters ?? zeroSpendCounters();
  const proofInput: CustomRingVelocityProofInput = Object.freeze({
    windowSlots: facts.windowSlots,
    rows: Object.freeze(rows.map((row) => Object.freeze({ ...row }))),
    ringId: hashBytes(decodeAddress(movement.ringProgramId)) as Bytes32,
    namespaceOwnerHash: ringNamespaceOwnerHash(facts.namespace),
    windowIndex: facts.windowIndex,
    approvalRequired,
    record: Object.freeze({
      version: spentRecord.version,
      window: spentRecord.window,
      commitment: spentRecord.countersCommitment,
      salt: opened.salt,
      assets: Object.freeze([...opened.assets]),
      spent: Object.freeze([...opened.spent]),
      nextSalt,
    }),
  });

  return Object.freeze({
    shape,
    nextNullifier: nextHashes.nullifier,
    recordInput,
    recordOutput,
    recordMessage: {
      viewTag: spendRecordMessageTag(namespace),
      data: encodeSpendRecord(successor),
    },
    countersSeal: sealedSpendCounters(nextCounters, namespace),
    proofInput,
    approvalRequired,
  });
}

/** Charges one transfer without a persistent spend record. */
export function chargeRows(
  input: Readonly<{
    movement: RingMovement;
    rows: readonly CustomRingVelocityRow[];
    namespaceOwnerHash: Bytes32;
  }>,
): CustomRingVelocityProofInput {
  let approvalRequired = false;
  const kept: CustomRingVelocityRow[] = [];
  for (const row of input.rows) {
    const outflow = senderOutflow(input.movement, row.asset);
    if (row.cap !== 0n && outflow > row.cap) {
      throw new RingError("RING_VELOCITY_CAP_EXCEEDED", {
        details: { asset: row.asset, cap: row.cap, spent: outflow },
      });
    }
    if (row.cosignAbove !== 0n && outflow > row.cosignAbove) approvalRequired = true;
    kept.push(Object.freeze({ ...row }));
  }
  return Object.freeze({
    ...velocityProofInputOff({
      ringId: hashBytes(decodeAddress(input.movement.ringProgramId)) as Bytes32,
      namespaceOwnerHash: input.namespaceOwnerHash,
    }),
    rows: Object.freeze(kept),
    approvalRequired,
  });
}
